### PR TITLE
[Code][Gameplay] Camera and Enhanced Input System

### PR DESCRIPTION
**Pull request by DrakeHunterK7**
Contains:

- Cinemachine customizable camera
- Unique player controls for both Gamepad and KB/M

Twin Stick shooter controls for gamepad, and mouse-driven player rotation for KB/M

### CODE REVIEW

File: Project Bunny/Assets/Scripts/Input/StudentController.cs [hunk 1]
```diff
@@ -1,5 +1,8 @@
 using UnityEngine;
 using UnityEngine.InputSystem;
+using System.Collections;
+using Cinemachine;
```
Comment (George540) [inline Project Bunny/Assets/Scripts/Input/StudentController.cs:4]: Remove unused variables and packages

File: Project Bunny/Assets/Scripts/Input/StudentController.cs [hunk 2]
```diff
@@ -8,19 +11,39 @@ public class StudentController : MonoBehaviour
...
+        [SerializeField] private Camera _playerCamera;
+        [SerializeField] private GameObject _playerModel;
+
+        [Header("Player")]
+        [SerializeField] private float _turnRate;
+
+        [Header("Camera")]
+
+        [SerializeField] 
+        [Range(0.0f, 90f)] private float _cameraAngle;
```
Comment (George540) [inline Project Bunny/Assets/Scripts/Input/StudentController.cs:23]: Can these be changed from the Cinemachine component? If so, no need to have them here

File: Project Bunny/Assets/Scripts/Input/StudentController.cs [hunk 3]
```diff
@@ -36,26 +59,65 @@ public void OnMove(InputAction.CallbackContext value)
...
+                        if(inputRotationL.magnitude > 0.1f)
+                        {
+                            Vector3 stickDirection = _playerCamera.transform.position + (new Vector3(inputRotationL.x, 0, inputRotationL.y) * 500f);
+                            Vector3 stickPosition = _characterController.transform.position + stickDirection;
+                            finalRotation = Quaternion.LookRotation(stickPosition);
+                        }
+                        
+                    }
+
+                    Debug.Log(finalRotation);
```
Comment (George540) [inline Project Bunny/Assets/Scripts/Input/StudentController.cs:98]: Remove debug statements. Unnecessary cluttering of console. Only use them for test purposes

File: Project Bunny/Assets/Scripts/Input/StudentController.cs [hunk 4]
```diff
@@ -36,26 +59,65 @@ public void OnMove(InputAction.CallbackContext value)
...
+            if(gamepad == null)
+            {
+                var mousePosition = Mouse.current.position.ReadValue();
+                var rotation = MousePosToRotationInput(mousePosition);
+                Debug.Log(rotation);
+                _playerModel.transform.rotation = Quaternion.Euler(0f, rotation, 0f);
+            }
+            else
+            {
+                if (_playerCamera is { })
```
Comment (George540) [inline Project Bunny/Assets/Scripts/Input/StudentController.cs:74]: Invert if statement. Rider can do that automatically if you select the suggestion button on the left

File: Project Bunny/Assets/Scripts/Input/StudentController.cs [hunk 5]
```diff
@@ -36,26 +59,65 @@ public void OnMove(InputAction.CallbackContext value)
...
+                            Vector3 stickDirection = _playerCamera.transform.position + (new Vector3(inputRotationL.x, 0, inputRotationL.y) * 500f);
+                            Vector3 stickPosition = _characterController.transform.position + stickDirection;
+                            finalRotation = Quaternion.LookRotation(stickPosition);
+                        }
+                        
+                    }
+
+                    Debug.Log(finalRotation);
+
+                    if(inputRotationR.magnitude > 0.1f || inputRotationL.magnitude > 0.1f)
```
Comment (George540) [inline Project Bunny/Assets/Scripts/Input/StudentController.cs:100]: Put if statement in brackets...always. Only one liner if statement allowed are the ones Rider suggests, which are usually the invert if statements. Also, you did not ident this right. This is why ALWAYS put if statements on brackets. Read code standards again

File: Project Bunny/Assets/Scripts/Input/StudentController.cs [hunk 6]
```diff
@@ -36,26 +59,65 @@ public void OnMove(InputAction.CallbackContext value)
...
+                var mousePosition = Mouse.current.position.ReadValue();
+                var rotation = MousePosToRotationInput(mousePosition);
+                Debug.Log(rotation);
+                _playerModel.transform.rotation = Quaternion.Euler(0f, rotation, 0f);
+            }
+            else
+            {
+                if (_playerCamera is { })
+                {
+                    Vector2 inputRotationR = gamepad.rightStick.ReadValue();
```
Comment (George540) [inline Project Bunny/Assets/Scripts/Input/StudentController.cs:76]: Use var type for local variables within methods

File: Project Bunny/Assets/Scripts/Input/StudentController.cs [hunk 7]
```diff
@@ -36,26 +59,65 @@ public void OnMove(InputAction.CallbackContext value)
...
         {
-            var mousePosition = Mouse.current.position.ReadValue();
-            var rotation = MousePosToRotationInput(mousePosition);
-            transform.rotation = Quaternion.Euler(0f, rotation, 0f);
+            var gamepad = Gamepad.current;
+            if(gamepad == null)
+            {
+                var mousePosition = Mouse.current.position.ReadValue();
+                var rotation = MousePosToRotationInput(mousePosition);
+                Debug.Log(rotation);
```
Comment (George540) [inline Project Bunny/Assets/Scripts/Input/StudentController.cs:69]: No debugs allowed

**Review by George540 — CHANGES_REQUESTED** [review]
A lot of coding standard changes are required

File: Project Bunny/Assets/Scripts/Input/StudentController.cs [hunk 8]
```diff
@@ -36,26 +59,65 @@ public void OnMove(InputAction.CallbackContext value)
...
+                    {
+                        Vector3 stickDirection = new Vector3(_playerCamera.transform.position.x, 0, _playerCamera.transform.position.y) + (new Vector3(inputRotationR.x, 0, inputRotationR.y) * 500f);
+                        Vector3 stickPosition = new Vector3(_characterController.transform.position.x, 0, _characterController.transform.position.y) + stickDirection;
+                        finalRotation = Quaternion.LookRotation(stickPosition);
+                    }
+                    else
+                    {
+                        if(inputRotationL.magnitude > 0.1f)
+                        {
+                            Vector3 stickDirection = _playerCamera.transform.position + (new Vector3(inputRotationL.x, 0, inputRotationL.y) * 500f);
```
Comment (George540) [inline Project Bunny/Assets/Scripts/Input/StudentController.cs:91]: Put var for these local variables again. Everything declared within a method are declared with var

File: Project Bunny/Assets/Scripts/Input/StudentController.cs (hunk 3 above)
Comment (DrakeHunterK7) [reply Project Bunny/Assets/Scripts/Input/StudentController.cs]: Donezo

File: Project Bunny/Assets/Scripts/Input/StudentController.cs (hunk 2 above)
Comment (DrakeHunterK7) [reply Project Bunny/Assets/Scripts/Input/StudentController.cs]: Everything about the camera can be changed in the Cinemachine component itself, so this is definitely useless here. Good catch!

File: Project Bunny/Assets/Scripts/Input/StudentController.cs (hunk 4 above)
Comment (DrakeHunterK7) [reply Project Bunny/Assets/Scripts/Input/StudentController.cs]: Done, downloaded Rider and fixed this.

File: Project Bunny/Assets/Scripts/Input/StudentController.cs (hunk 6 above)
Comment (DrakeHunterK7) [reply Project Bunny/Assets/Scripts/Input/StudentController.cs]: Done 2: Return of the Done

**Comment by DrakeHunterK7** [comment]
> A lot of coding standard changes are required

Agreed and thank you very much for calling me out on these. Definitely will improve going forward. Thanks!

File: Project Bunny/Assets/Scripts/Input/StudentController.cs (hunk 8 above)
Comment (DrakeHunterK7) [reply Project Bunny/Assets/Scripts/Input/StudentController.cs]: Done

**Review by Cappic-01 — APPROVED** [review]
Coding conventions corrected and Cinemachine looking good!

**Review by George540 — APPROVED** [review]
Very good, yum yum

**Review by George540 — APPROVED** [review]
Cool cool

**Review by Cappic-01 — APPROVED** [review]
Btw if using Summaries, don't forget to also describe any parameter and what the function returns. There are fields generated automatically just for that in the Summary block.

**Pull request merged**